### PR TITLE
LL1 parser fix

### LINKED ISSUE #32: FIX THE PARSER
# BUG REPORT: FIX THE PARSER

**OS**: Ubuntu (LTS) 20.04 or macOS 10.14+ (please specify version).
**Compiler**: GCC 7.0+ or Clang 8.0+. 
**CMake Profile**: `all`

### grammar.g
```
## This is the grammar file for JuCC
## Edit this file to make changes to the parsing grammar
## Epsilon is represented by special string EPSILON

## Terminals
%terminals
else float if int void
( ) { } * + - / % ,
<< >> < > <= >= = == != ;
identifier integer_constant float_constant
main cin cout
%

## Non Terminals
%non_terminals
<primary_expression> <constant> <unary_operator> <unary_expression>
<type_specifier> <multiplicative_expression> <additive_expression>
<shift_expression> <relational_expression> <equality_expression>
<assignment_expression> <expression>
<declaration> <init_declarator_list> <init_declarator>
<initializer> <declarator> <direct_declarator>
<statement> <compound_statement> <block_item_list> <block_item>
<expression_statement> <selection_statement> <program>
%

## Start Symbol
%start
<program>
%

## Grammar for the language
%rules
## Expressions
<primary_expression> : identifier
<primary_expression> : <constant>
<primary_expression> : ( <expression> )
<constant> : integer_constant
<constant> : float_constant
<unary_operator> : +
<unary_operator> : -
<unary_expression> : <primary_expression>
<unary_expression> : <unary_operator> <primary_expression>
<multiplicative_expression> : <unary_expression>
<multiplicative_expression> : <multiplicative_expression> * <unary_expression>
<multiplicative_expression> : <multiplicative_expression> / <unary_expression>
<multiplicative_expression> : <multiplicative_expression> % <unary_expression>
<additive_expression> : <multiplicative_expression>
<additive_expression> : <additive_expression> + <multiplicative_expression>
<additive_expression> : <additive_expression> - <multiplicative_expression>
<shift_expression> : <additive_expression>
<shift_expression> : cin >> <additive_expression>
<shift_expression> : cout << <additive_expression>
<shift_expression> : <shift_expression> << <additive_expression>
<shift_expression> : <shift_expression> >> <additive_expression>
<relational_expression> : <shift_expression>
<relational_expression> : <relational_expression> < <shift_expression>
<relational_expression> : <relational_expression> > <shift_expression>
<relational_expression> : <relational_expression> <= <shift_expression>
<relational_expression> : <relational_expression> >= <shift_expression>
<equality_expression> : <relational_expression>
<equality_expression> : <equality_expression> == <relational_expression>
<equality_expression> : <equality_expression> != <relational_expression>
<assignment_expression> : <equality_expression>
<assignment_expression> : <assignment_expression> = <equality_expression>
<expression> : <assignment_expression>

## Declarations
<declaration> : <type_specifier> <init_declarator_list> ;
<init_declarator_list> : <init_declarator>
<init_declarator_list> : <init_declarator_list> , <init_declarator>
<init_declarator_list> : EPSILON
<init_declarator> : <declarator>
<init_declarator> : <declarator> = <initializer>
<type_specifier> : void
<type_specifier> : int
<type_specifier> : float
<declarator> : <direct_declarator>
<direct_declarator> : identifier
<direct_declarator> : ( <declarator> )
<initializer> : <assignment_expression>

## Statements
<statement> : <compound_statement>
<statement> : <expression_statement>
<statement> : <selection_statement>
<compound_statement> : { <block_item_list> }
<block_item_list> : <block_item>
<block_item_list> : <block_item> <block_item_list> 
<block_item_list> : EPSILON
<block_item> : <declaration>
<block_item> : <statement>
<expression_statement> : <expression> ;
<expression_statement> : ;
<selection_statement> : if ( <expression> ) <compound_statement>
<selection_statement> : if ( <expression> ) <compound_statement> else <compound_statement>

## Main
<program> : <type_specifier> main ( ) <compound_statement>
%
```
### intest1.cc
```
int main() {
	int x = 1 + 2 + 3 + 4;
	if 1;
}

```

### intest2.cc
```
int

```

## Expected output:
Some error :anger: 

## Current Output:
```
Segmentation fault (core dumped)
```

### PR DESCRIPTION
**Pull request by bisakhmondal**
# New edge case
closes #32

### CODE REVIEW

File: src/main/main.cpp [hunk 1]
```diff
@@ -94,9 +94,19 @@ auto main(int argc, char *argv[]) -> int {
...
   int token;
+
   while ((token = lexer.GetToken(ifs)) != jucc::lexer::TOK_EOF) {
     input_tokens.emplace_back(jucc::lexer::Lexer::GetTokenType(token));
   }
+  std::vector<std::string> errors = lexer.GetUndeclaredSymbolErrors();
+  errors.insert(errors.end(), lexer.GetDuplicateSymbolErrors().begin(), lexer.GetDuplicateSymbolErrors().end());
+
+  if (!errors.empty()) {
+    for (auto &e : errors) {
```
Comment (noob77777) [inline src/main/main.cpp:106]: std::cout << "jucc: "; between line 104 and 105

File: test/parser/grammar.g [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+<block_item> : <declaration>
+<block_item> : <statement>
+<expression_statement> : <expression> ;
+<expression_statement> : ;
+<selection_statement> : if ( <expression> ) <compound_statement>
+<selection_statement> : if ( <expression> ) <compound_statement> else <compound_statement>
+
+## Main
+<program> : <type_specifier> main ( ) <compound_statement>
+%
```
Comment (noob77777) [inline test/parser/grammar.g]: add newline

**Review by noob77777 — CHANGES_REQUESTED** [review]

File: src/main/main.cpp [hunk 3]
```diff
@@ -94,9 +94,19 @@ auto main(int argc, char *argv[]) -> int {
...
+      std::cout << e << "\n";
```
Comment (noob77777) [inline src/main/main.cpp]: use '\n' for uniformity

**Comment by TheSYNcoder** [comment]
Wow all changes done,
That comment worked.

**Comment by TheSYNcoder** [comment]
Merge

File: src/main/main.cpp (hunk 1 above)
Comment (bisakhmondal) [reply src/main/main.cpp:106]: done

File: test/parser/grammar.g (hunk 2 above)
Comment (bisakhmondal) [reply test/parser/grammar.g]: done

File: src/main/main.cpp (hunk 3 above)
Comment (bisakhmondal) [reply src/main/main.cpp]: done

**Review by noob77777 — APPROVED** [review]

**Pull request merged**